### PR TITLE
Improve Dense HLL merge performance

### PR DESCRIPTION
**Pull request by martint**

### CODE REVIEW

File: stats/src/test/java/io/airlift/stats/cardinality/BenchmarkDenseHll.java [hunk 1]
```diff
@@ -63,6 +72,27 @@ public void initialize()
...
+    public static class MergeData
+    {
+        public DenseHll base;
+        public DenseHll toMerge;
+
+        @Setup(Level.Iteration)
+        public void initialize()
+        {
+            base = new DenseHll(12);
+            for (int i = 0; i < 1_000_000; i++) {
```
Comment (sopel39) [inline stats/src/test/java/io/airlift/stats/cardinality/BenchmarkDenseHll.java]: Is it important that here are `1_000_000` and below `100` only? Should that be parametrized?

File: stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java [hunk 2]
```diff
@@ -249,16 +249,14 @@ public void insert(int bucket, int value)
...
-            if (!setOverflow(bucket, overflow)) {
-                // grow overflows arrays if necessary
-                overflowBuckets = Ints.ensureCapacity(overflowBuckets, overflows + 1, OVERFLOW_GROW_INCREMENT);
-                overflowValues = Bytes.ensureCapacity(overflowValues, overflows + 1, OVERFLOW_GROW_INCREMENT);
-
-                overflowBuckets[overflows] = bucket;
-                overflowValues[overflows] = (byte) overflow;
-                overflows++;
+            int overflowEntry = findOverflowEntry(bucket);
+            if (overflowEntry != -1) {
```
Comment (sopel39) [inline stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java]: Why this makes it faster? It introduces branching

File: stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java [hunk 3]
```diff
@@ -272,28 +270,24 @@ public void insert(int bucket, int value)
         }
     }
 
-    private int getOverflow(int bucket)
+    private int findOverflowEntry(int bucket)
     {
         for (int i = 0; i < overflows; i++) {
             if (overflowBuckets[i] == bucket) {
```
Comment (sopel39) [inline stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java]: should that be some hash map instead?

**Comment by sopel39** [comment]
@martint did you profile HLL after your changes? What takes the most time?

File: stats/src/main/java/io/airlift/stats/cardinality/BucketListener.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+/*
```
Comment (sopel39) [inline stats/src/main/java/io/airlift/stats/cardinality/BucketListener.java]: Do you have results from before?

File: stats/src/test/java/io/airlift/stats/cardinality/BenchmarkDenseHll.java (hunk 1 above)
Comment (martint) [reply stats/src/test/java/io/airlift/stats/cardinality/BenchmarkDenseHll.java]: No special meaning other than having a "large" and a "small" set for the comparison with sparse hll in the next commit to be meaningful (sparse can store only a small set).

File: stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java (hunk 2 above)
Comment (martint) [reply stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java]: `delta > MAX_DELTA` is statistically unlikely, so this branch doesn't hit often. In any case, the old code already had a branch `if (!setOverflow(...))`. This just moves things around a bit.

File: stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java (hunk 3 above)
Comment (martint) [reply stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java]: Possibly, but overflows are statistically infrequent and there's usually only a few of them. Scanning a few items may be faster than computing a hashcode.

File: stats/src/main/java/io/airlift/stats/cardinality/BucketListener.java (hunk 4 above)
Comment (martint) [reply stats/src/main/java/io/airlift/stats/cardinality/BucketListener.java]: Before, we'd convert the sparse HLL into a dense HLL and then merge it. So the performance is approximately what the first benchmark showed (there's some extra cost in the conversion that we're not measuring):

`21.557 ± 1.116 us/op` for current implementation,
`11.677 ± 0.788  us/op` after Dense HLL merging is optimized,
`0.320 ± 0.013 us/op` after dedicated merge implementation for Dense vs Sparse

File: stats/src/main/java/io/airlift/stats/cardinality/HyperLogLog.java [hunk 5]
```diff
@@ -84,6 +84,9 @@ public void mergeWith(HyperLogLog other)
             ((SparseHll) instance).mergeWith((SparseHll) other.instance);
             instance = makeDenseIfNecessary((SparseHll) instance);
         }
+        else if (instance instanceof DenseHll && other.instance instanceof SparseHll) {
```
Comment (dain) [inline stats/src/main/java/io/airlift/stats/cardinality/HyperLogLog.java]: What about the other way around?

Comment (martint) [reply stats/src/main/java/io/airlift/stats/cardinality/HyperLogLog.java]: In that case, the sparse needs to be converted to dense before the merge can take place. That’s handled by the fall-through case

File: stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java (hunk 3 above)
Comment (sopel39) [reply stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java]: Right, could you just use array with number of buckets length? Would that help?

Comment (martint) [reply stats/src/main/java/io/airlift/stats/cardinality/DenseHll.java]: That would be a lot of wasted space. Out of, say, 4096 buckets, there might only < 10 that overflow at any given point.

**Pull request merged**